File: Algo/question/074SearchA2DMatrix.cpp
//1. Binary search. Just in matrix. O(lg(m+n)).
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        if(m==0) return false;
        int n = matrix[0].size();
        if(n==0) return false;        
        
        int l = 0, r = n*m - 1;
        while(l<=r){
            int mid = (l+r)/2;
            int value = matrix[mid/n][mid%n];
            
            if(value<target) l = mid+1;
            else if(value>target) r = mid - 1;
            else {return true;}
        }
        
        return false;
    }
};

File: Algo/Leetcode/074SearchA2DMatrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        n = len(matrix)
        m = len(matrix[0])
        i = 0
        j = n * m - 1
        while i <= j:
            mid = (i + j) / 2
            u, v = flat_to_uv(mid, n, m)
            if matrix[u][v] == target:
                return True
            elif matrix[u][v] > target:
                j = mid - 1
            else:
                i = mid + 1

        return False

def flat_to_uv(v, n, m):
    return v / m, v % m